upcloud/resource_upcloud_objectstorage.go

```diff
@@ -0,0 +1,205 @@
+				Required:     true,
+				ValidateFunc: validation.StringLenBetween(8, 255),
+			},
+			"zone": {
+				Description: "The zone in which the object storage bucket will be created",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			"name": {
+				Description: "The name of the object storage bucket to be created",
+				Optional:    true,
+				Type:        schema.TypeString,
+			},
+			"description": {
+				Description: "The description of the object storage bucket to be created",
+				Optional:    true,
+				Type:        schema.TypeString,
+				ValidateFunc: validation.StringLenBetween(1, 255),
+			},
+			"url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"used_space": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"created": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceObjectStorageCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var (
+	    diags diag.Diagnostics
+		req request.CreateObjectStorageRequest
+	)
+
+	client := m.(*service.Service)
+
+	req.Size = d.Get("size").(int)
+	req.Zone = d.Get("zone").(string)
+	req.AccessKey = d.Get("access_key").(string)
+	req.SecretKey = d.Get("secret_key").(string)
+
+	if v, ok := d.GetOk("name"); ok {
+		req.Name = v.(string)
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		req.Description = v.(string)
+	}
+
+	objStorage, err := client.CreateObjectStorage(&req)
+	if err != nil {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Unable to create object storage",
+			Detail:   err.Error(),
+		})
+		return diags
+	}
+
+	d.SetId(objStorage.UUID)
+
+	copyObjectStorageDetails(objStorage, d)
+
+	return diags
+}
+
+func resourceObjectStorageRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*service.Service)
+
+	// Warning or errors can be collected in a slice type
+	var diags diag.Diagnostics
+
+	uuid := d.Id()
+
+	objectDetails, err := client.GetObjectStorageDetails(&request.GetObjectStorageDetailsRequest{
+		UUID: uuid,
+	})
+
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	copyObjectStorageDetails(objectDetails, d)
+
+	return diags
+}
+
+func resourceObjectStorageUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*service.Service)
+
+	if d.HasChanges([]string{"size", "access_key", "secret_key", "description"}...) {
+		req := request.ModifyObjectStorageRequest{
+			UUID: d.Id(),
+		}
+
+		if d.HasChange("size") {
+			req.Size = d.Get("size").(int)
+		}
+
+		if d.HasChange("access_key") {
+			req.AccessKey = d.Get("access_key").(string)
+		}
+
+		if d.HasChange("secret_key") {
+			req.SecretKey = d.Get("secret_key").(string)
+		}
+
+		if d.HasChange("description") {
+			req.Description = d.Get("description").(string)
+		}
```

Comment:
we might want to populate the values depending on whether they have changed or not (in case someone pops omit empty away from to go api for some reason)